src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+                {
+                    var memberReference = (IMemberReferenceOperation)operationContext.Operation;
+                    if (IsCandidateSymbol(memberReference.Member))
+                    {
+                        // Get the value usage info.
+                        var valueUsageInfo = memberReference.GetValueUsageInfo();
+
+                        // Special case for VB
+                        // Compound assigment is a statement in VB that does not return a value.
+                        // So, we treat this usage as a write-only usage.
+                        if (memberReference.Language == LanguageNames.VisualBasic &&
+                            memberReference.Parent is ICompoundAssignmentOperation compoundAssignment &&
+                            compoundAssignment.Target == memberReference)
+                        {
+                            valueUsageInfo = ValueUsageInfo.Write;
+                        }
+
+                        OnSymbolUsage(memberReference.Member, valueUsageInfo);
+                    }
+                }
+
+                void AnalyzeInvocationOperation(OperationAnalysisContext operationContext)
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+                    if (IsCandidateSymbol(invocation.TargetMethod))
+                    {
+                        OnSymbolUsage(invocation.TargetMethod, ValueUsageInfo.Read);
+                    }
+                }
+
+                void OnSymbolEnd(SymbolAnalysisContext symbolEndContext)
+                {
+                    // Report diagnostics for unused candidate members.
+                    var members = ((INamedTypeSymbol)symbolEndContext.Symbol).GetMembers();
+                    foreach (var member in members)
+                    {
+                        // Check if the underlying member is neither read nor a readable reference to the member is taken.
+                        if (symbolValueUsageStateMap.TryRemove(member, out var valueUsageInfo) &&
+                            (valueUsageInfo & (ValueUsageInfo.Read | ValueUsageInfo.ReadableRef)) == 0)
+                        {
+                            Debug.Assert(IsCandidateSymbol(member));
+
+                            var option = GetCodeStyleOption(member, symbolEndContext);
+                            if (option != null && option.Value)
+                            {
+                                // Give appropriate message based on usage:
+                                //  1. No read or writes:   Type '{0}' has an unused private member '{1}' which can be removed.
+                                //  2. Only writes:         Type '{0}' has a private member '{1}' which can be removed as the value assigned to it is never used.
+
+                                var rule = (valueUsageInfo & (ValueUsageInfo.Write | ValueUsageInfo.WritableRef)) != 0 ?
```

Comment:
i've seen this pattern repeat all oer the PR.  please create some helper methods/extensions for asking these questions.  There's no need for everyone to have to think through all the bitwise operations each time we use this type.